### PR TITLE
Reland optimized-compiler-builtins config

### LINKED ISSUE #101172: bootstrap cannot check library/std for windows from macOS (and possibly others)
I often write patches for `std::sys::windows` while not using windows. It would be nice if `./x.py check --target x86_64-pc-windows-msvc library/std` worked, but instead it does not -- I get an error like https://gist.github.com/thomcc/7c889b3a1306922f7eda28f45f1d304e (this is from a build where `aarch64-apple-darwin` is the host).

As a result, to check changes to std::sys::windows while not on windows, I'm forced to a full stage1 build, and then in some other workspace run `cargo +stage1 check -Zbuild-std --target x86_64-pc-windows-msvc` (where `+stage1` is a linked toolchain).  This works, but is tedious/time-consuming.

CC @jyn514 who suggested I file this issue, and suggested that it "looks like we're handling the linker arguments wrong when cross compiling".

### LINKED ISSUE #102560: Decouple `is_rust_llvm` from whether LLVM is in-tree or not
The point of `is_rust_llvm` is to know whether it contains Rust patches. In #101072 @tmandry added a config option so you can use a vanilla upstream LLVM without affecting other parts of how bootstrap works, this breaks that by re-coupling the idea of "is patched" to "is in tree".

_Originally posted by @tmandry in https://github.com/rust-lang/rust/pull/101833#discussion_r984872396_

### PR DESCRIPTION
**Pull request by onur-ozkan**
Copy of #102579 PR.

From #102579:

> No concerns on my side. Currently, Jyn isn't actively working on the project. I will close this PR; open another one to cherry-pick the commits, resolve conflicts, and then r+ it.

> Fixes https://github.com/rust-lang/rust/issues/102560. Fixes https://github.com/rust-lang/rust/issues/101172. Helps with https://github.com/rust-lang/rust/issues/105065 (although there's some weirdness there - it's still broken when optimized-compiler-builtins is set to true).

Fixes https://github.com/rust-lang/rust/issues/102560. Fixes https://github.com/rust-lang/rust/issues/101172. Helps with https://github.com/rust-lang/rust/issues/105065

r? ghost

### CODE REVIEW

**Comment by matthiaskrgr** [comment]
@bors r- rollup=never
looks like it failed in https://github.com/rust-lang/rust/pulls#issuecomment-1877189661

File: src/bootstrap/src/utils/change_tracker.rs [hunk 1]
```diff
@@ -106,4 +106,9 @@ pub const CONFIG_CHANGE_HISTORY: &[ChangeInfo] = &[
         severity: ChangeSeverity::Info,
         summary: "The dist.missing-tools config option was deprecated, as it was unused. If you are using it, remove it from your config, it will be removed soon.",
     },
+    ChangeInfo {
+        change_id: 102579,
```
Comment (kadiwa4) [inline src/bootstrap/src/utils/change_tracker.rs:110]: you didn't update the PR number

Comment (onur-ozkan) [reply src/bootstrap/src/utils/change_tracker.rs:110]: It's the actual PR number of the implementation.

File: src/ci/run.sh [hunk 2]
```diff
@@ -80,6 +80,15 @@ fi
 # space required for CI artifacts.
 RUST_CONFIGURE_ARGS="$RUST_CONFIGURE_ARGS --dist-compression-formats=xz"
 
+# Enable the `c` feature for compiler_builtins, but only when the `compiler-rt` source is available
+# (to avoid spending a lot of time cloning llvm)
+if [ "$EXTERNAL_LLVM" = "" ]; then
+  RUST_CONFIGURE_ARGS="$RUST_CONFIGURE_ARGS --set build.optimized-compiler-builtins"
+elif [ "$DEPLOY$DEPLOY_ALT" = "1" ]; then
+    echo "error: dist builds should always use optimized compiler-rt!" >&2
```
[the remark this answers is not in the document]
Comment (nikic) [reply src/ci/run.sh:88]: Does dist-various-2 really use an external LLVM? I don't see llvm-root/llvm-config options in the Dockerfile.

[the remark this answers is not in the document]
Comment (onur-ozkan) [reply src/ci/run.sh:88]: It doesn't use, my bad. I got confused with "ENV EXTERNAL_LLVM 1"

**Pull request merged**